src/librustc/lint/builtin.rs

```diff
@@ -2083,6 +2097,23 @@ impl LintPass for PrivateNoMangleFns {
                     cx.span_lint(PRIVATE_NO_MANGLE_FNS, it.span, msg.as_slice());
                 }
             },
+            ast::ItemStatic(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") &&
+                       !cx.exported_items.contains(&it.id) {
+                    let msg = format!("static {} is marked #[no_mangle], but not exported",
+                                      it.ident);
+                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
+                }
+            },
+            ast::ItemConst(..) => {
+                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
+                    // Const items do not refer to a particular location in memory, and therefore
+                    // don't have anything to attach a symbol to
+                    let msg = "const items should never be #[no_mangle]";
+                    cx.span_lint(NO_MANGLE_CONST_ITEMS, it.span, msg);
+                    cx.sess().span_help(it.span, "maybe you wanted this to be a pub static");
```

Comment:
I'm not sure that's a huge issue in this case. I'm not aware of any time
you would actually want this.

On Tuesday, February 3, 2015, Keegan McAllister <notifications@github.com>
wrote:

> In src/librustc/lint/builtin.rs
> <https://github.com/rust-lang/rust/pull/21881#discussion_r24038887>:
>
> > @@ -2083,6 +2097,23 @@ impl LintPass for PrivateNoMangleFns {
> >                      cx.span_lint(PRIVATE_NO_MANGLE_FNS, it.span, msg.as_slice());
> >                  }
> >              },
> > +            ast::ItemStatic(..) => {
> > +                if attr::contains_name(it.attrs.as_slice(), "no_mangle") &&
> > +                       !cx.exported_items.contains(&it.id) {
> > +                    let msg = format!("static {} is marked #[no_mangle], but not exported",
> > +                                      it.ident);
> > +                    cx.span_lint(PRIVATE_NO_MANGLE_STATICS, it.span, msg.as_slice());
> > +                }
> > +            },
> > +            ast::ItemConst(..) => {
> > +                if attr::contains_name(it.attrs.as_slice(), "no_mangle") {
> > +                    // Const items do not refer to a particular location in memory, and therefore
> > +                    // don't have anything to attach a symbol to
> > +                    let msg = "const items should never be #[no_mangle]";
> > +                    cx.span_lint(NO_MANGLE_CONST_ITEMS, it.span, msg);
> > +                    cx.sess().span_help(it.span, "maybe you wanted this to be a pub static");
>
> Unfortunately the help message will print even with -A
> no-mangle-const-items, see #19668
> <https://github.com/rust-lang/rust/issues/19668>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rust-lang/rust/pull/21881/files#r24038887>.
>